analyses/cell-type-consensus/scripts/05-generate-validation-markers.R

```diff
@@ -0,0 +1,98 @@
+# define output file 
+validation_markers_file <- file.path("references", "validation-markers.tsv")
+
+# Prep references --------------------------------------------------------------
+# Read in consensus groups
+consensus_groups_df <- readr::read_tsv(consensus_validation_file) |>
+  dplyr::select(validation_group_annotation, validation_group_ontology) |> 
+  unique()
+
+# get list of all ontology IDs that we care about 
+group_ontology_ids <- consensus_groups_df |> 
+  dplyr::pull(validation_group_ontology) |> 
+  unique()
+
+# read in cell marker and filter to relevant ontology IDs
+cell_marker_df <- readxl::read_xlsx(cellmarker_file) |> 
+  dplyr::mutate(
+    cellontology_id = stringr::str_replace(cellontology_id, "_", ":")
+  ) |> 
+  dplyr::filter(cellontology_id %in% group_ontology_ids) |> 
+  # rename for later joining
+  dplyr::select(validation_group_ontology = cellontology_id, gene_symbol = Symbol, tissue_type) |> 
+  # get rid of any thing that doesn't have a gene symbol 
+  tidyr::drop_na() |> 
+  unique() # only want one instance of each cell type/gene combo per tissue type 
+
+# get ensembl ids for genes 
+mapped_ids <-AnnotationDbi::mapIds(
+  org.Hs.eg.db,
+  keys = unique(cell_marker_df$gene_symbol),
+  keytype = "SYMBOL",
+  column = "ENSEMBL",
+  multiVals = "first"
+) |>
+  tibble::enframe(name = 'gene_symbol',
+                  value = 'ensembl_gene_id')
+
+# Get top markers -------------------------------------------------------------
+
+# get total number of tissues for each cell type 
+total_tissues_df <- cell_marker_df |> 
+  dplyr::select(validation_group_ontology, tissue_type) |> 
+  unique() |>
+  dplyr::count(validation_group_ontology, name = "celltype_total_tissues")
+
+# create data frame with top 10 marker genes per cell type 
+# ranking is based on the percentage of tissues the gene shows up in for a given cell type  
+top_markers_df <- cell_marker_df |> 
+  # first count how many tissues each gene shows up in for each cell type 
+  dplyr::count(validation_group_ontology, gene_symbol, sort = TRUE, name = "number_of_tissues") |> 
+  # add total possible tissue types 
+  dplyr::left_join(total_tissues_df) |> 
+  dplyr::mutate(
+    # get the total percent of tissues that have that marker gene in that cell type 
+    percent_tissues = round((number_of_tissues/celltype_total_tissues) * 100, 2)
+  ) |> 
+  # get the top 10 for each group, if there's a tie all will be saved
+  dplyr::group_by(validation_group_ontology) |> 
+  dplyr::top_n(10, percent_tissues) |> 
+  # arrange genes by cell type and percentage of tissues 
+  dplyr::arrange(validation_group_ontology, desc(percent_tissues)) |> 
+  # bring in ensembl ids
+  dplyr::left_join(mapped_ids, by = c("gene_symbol")) |>
+  # match to human readable name for ontology group 
+  dplyr::left_join(consensus_groups_df, by = c("validation_group_ontology")) |> 
+  dplyr::relocate(
+    validation_group_annotation, ensembl_gene_id, .after = validation_group_ontology
+  )
```

Comment:
Should we be concerned about the specificity of marker genes, too? This only considers how often a gene shows up for a cell type across tissues, but what if the same gene shows up for many cell types? I don't know that we want that for our purposes. CellMarker might be built with this in mind (I don't recall), or we can ameliorate my concern by demonstrating this doesn't come up in practice.

Reply:
To address this I added a column for `gene_observed_count` which notes how many times that gene shows up in the table and therefore how many cell types it's found in. For now I didn't do any further filtering on this column, but we may want to in the future when we go to actually make the figure. 

For some additional context, most cell types have at least 2 marker genes that are specific to that cell type. I included a table with the total counts of unique marker genes per group below. 
<img width="258" alt="Screenshot 2025-02-17 at 10 25 04 AM" src="https://github.com/user-attachments/assets/75b5926e-169c-404a-bfd6-612b317ed0e4" />